_posts/2020-12-11-torch-0.2.0-released/torch-0.2.0-released.Rmd

```diff
@@ -0,0 +1,310 @@
+dataloaders behave like [{coro}](https://github.com/r-lib/coro) iterators. 
+This means that you can now use [{coro}](https://github.com/r-lib/coro)'s syntax 
+to loop trough the dataloaders:
+
+```{r}
+coro::loop(for(batch in par_dl) {
+  print(batch$shape)
+})
+```
+
+```
+[1] 5 1
+[1] 5 1
+```
+
+This is the first {torch} release including the multiworker
+dataloaders feature and you might run into edge cases when
+using it. Do let us know if you find any problem.
+
+## Initial JIT support
+
+Programs that make use of the {torch} package are inevitably
+R programs thus, they always need an R installation in order
+to execute.
+
+As of version 0.2.0, {torch} allows users to JIT *trace* 
+{torch} R functions into TorchScript. The nice thing about
+this is that TorchScript programs are easily serializable, 
+optimizable and they can be loaded in another program writen
+in PyTorch or LibTorch without requiring any R dependency.
+
+Suppose you have the following R function that takes a tensor
+and does a matrix multiplication with a fixed weight matrix and
+then adds a bias term:
+
+```{r}
+w <- torch_randn(10, 1)
+b <- torch_randn(1)
+fn <- function(x) {
+  a <- torch_mm(x, w)
+  a + b
+}
+```
+
+This function can be JIT traced into TorchScript with `jit_trace` by passing the function and example inputs: 
+
+```{r}
+x <- torch_ones(2, 10)
+tr_fn <- jit_trace(fn, x)
+tr_fn(x)
+```
+```
+torch_tensor
+-0.6880
+-0.6880
+[ CPUFloatType{2,1} ]
+```
+
+Now all {torch} operations that happened when computing the result of 
+this function were traced and transformed into a graph:
+
+```{r}
+tr_fn$graph
+```
+```
+graph(%0 : Float(2:10, 10:1, requires_grad=0, device=cpu)):
+  %1 : Float(10:1, 1:1, requires_grad=0, device=cpu) = prim::Constant[value=-0.3532  0.6490 -0.9255  0.9452 -1.2844  0.3011  0.4590 -0.2026 -1.2983  1.5800 [ CPUFloatType{10,1} ]]()
+  %2 : Float(2:1, 1:1, requires_grad=0, device=cpu) = aten::mm(%0, %1)
+  %3 : Float(1:1, requires_grad=0, device=cpu) = prim::Constant[value={-0.558343}]()
+  %4 : int = prim::Constant[value=1]()
+  %5 : Float(2:1, 1:1, requires_grad=0, device=cpu) = aten::add(%2, %3, %4)
+  return (%5)
+```
+
+The traced function can be serialized with `jit_save`:
+
+```{r}
+jit_save(tr_fn, "linear.pt")
+```
+
+It can be reloaded in R with `jit_load`, but it can also be reloaded in Python
+with `torch.jit.load`:
+
+```{python}
+import torch
+fn = torch.jit.load("linear.pt")
+fn(torch.ones(2, 10))
+```
+
+```
+tensor([[-0.6880],
+        [-0.6880]])
+```
+
+How cool is that?!
+
+This is just the initial support for the JIT in R. We will continue developing 
+this, specially, in the next version of {torch} we plan to support tracing `nn_modules` directly. Currently, you need to detach all parameters before 
+tracing them, see an example [here](https://github.com/mlverse/torch/blob/master/tests/testthat/test-trace.R#L27-L71). This will allow you also to take benefit of TorchScript to make your models
```

Comment:
```suggestion
tracing them; see an example [here](https://github.com/mlverse/torch/blob/master/tests/testthat/test-trace.R#L27-L71). This will allow you also to take benefit of TorchScript to make your models
```
```suggestion
tracing them, see an example [here](https://github.com/mlverse/torch/blob/master/tests/testthat/test-trace.R#L27-L71). This will also allow you to benefit from TorchScript to make your models
```